timer.r/scripts/paralell.r

```diff
@@ -0,0 +1,153 @@
+# Assignment 1:  
+library(tweedie) 
+library(ggplot2)
+library(tictoc)
+library(knitr)
+
+
+tic()
+simTweedieTest <-  
+  function(N){ 
+    t.test( 
+      rtweedie(N, mu=10000, phi=100, power=1.9), 
+      mu=10000 
+    )$p.value 
+  } 
+toc()
+
+
+# Assignment 2:  
+
+tic()
+MTweedieTests <-  
+  function(N,M,sig){ 
+    sum(replicate(M,simTweedieTest(N)) < sig)/M 
+  } 
+toc()
+
+# Assignment 3:  
+#Load package
+library(doParallel)
+
+# Register the number of cores to be used for parallel computation
+# This example uses 8 cores. 
+maxcores <- 8
+Cores <- min(parallel::detectCores(), maxcores)
+cl <- makeCluster(Cores)
+registerDoParallel(cl)
+
+df <-  
+  expand.grid( 
+    N = c(10,100,1000,5000, 10000), 
+    M = 1000, 
+    share_reject = NA) 
+
+# Export necessary functions and data to the cluster workers:
+clusterExport(cl, list("MTweedieTests", "simTweedieTest"))
+
+tic(paste0("Parallel loop for MTweedieTests, ", Cores, " cores"))
+
+for(i in 1:nrow(df)){ 
+  df$share_reject[i] <-  
+    MTweedieTests( 
+      N=df$N[i], 
+      M=df$M[i], 
+      sig=.05) 
+} 
```

Comment:
Nice try, but not completely correct. To use parallell computing you should have used the foreach() function. Look at the solution proposal below:

```r
# Next we use doParallel for the loop over the result df: 
maxcores <- configs$n_cores
Cores <- min(parallel::detectCores(), maxcores)
cl <- makeCluster(Cores)
registerDoParallel(cl)


res <-
  foreach(
    i = 1:nrow(df),
    .combine = 'rbind',
    .packages = c('magrittr', 'dplyr', 'tweedie')
  ) %dopar%
  tibble(
    N = df$N[i],
    M = df$M[i],
    share_reject =
      MTweedieTests(N = df$N[i],
                    M = df$M[i],
                    sig = configs$significance_level)
  )

# Now that we're done, we close off the clusters
stopCluster(cl)
```